Sources_v3/StreamChatUI/ChatChannel/ChatMessageAttachmentsView.swift

```diff
@@ -0,0 +1,116 @@
+    public var data: AttachmentListViewData<ExtraData>? {
+        didSet { updateContentIfNeeded() }
+    }
+
+    // MARK: - Subviews
+
+    public private(set) lazy var imageGallery = uiConfig
+        .messageList
+        .messageContentSubviews
+        .attachmentSubviews
+        .imageGallery
+        .init()
+        .withoutAutoresizingMaskConstraints
+
+    public private(set) lazy var fileList = uiConfig
+        .messageList
+        .messageContentSubviews
+        .attachmentSubviews
+        .fileAttachmentListView
+        .init()
+        .withoutAutoresizingMaskConstraints
+
+    private var layoutConstraints: [LayoutOptions: [NSLayoutConstraint]] = [:]
+
+    // MARK: - Overrides
+
+    override open func setUpLayout() {
+        addSubview(imageGallery)
+        addSubview(fileList)
+
+        layoutConstraints[.images] = [
+            imageGallery.leadingAnchor.constraint(equalTo: leadingAnchor),
+            imageGallery.trailingAnchor.constraint(equalTo: trailingAnchor),
+            imageGallery.topAnchor.constraint(equalTo: topAnchor),
+            imageGallery.bottomAnchor.constraint(equalTo: bottomAnchor)
+        ]
+
+        layoutConstraints[.files] = [
+            fileList.leadingAnchor.constraint(equalTo: leadingAnchor),
+            fileList.trailingAnchor.constraint(equalTo: trailingAnchor),
+            fileList.topAnchor.constraint(equalTo: topAnchor),
+            fileList.bottomAnchor.constraint(equalTo: bottomAnchor)
+        ]
+
+        layoutConstraints[[.images, .files]] = [
+            imageGallery.leadingAnchor.constraint(equalTo: leadingAnchor),
+            imageGallery.trailingAnchor.constraint(equalTo: trailingAnchor),
+            imageGallery.topAnchor.constraint(equalTo: topAnchor),
+            
+            fileList.leadingAnchor.constraint(equalTo: leadingAnchor),
+            fileList.trailingAnchor.constraint(equalTo: trailingAnchor),
+            fileList.topAnchor.constraint(equalTo: imageGallery.bottomAnchor),
+            fileList.bottomAnchor.constraint(equalTo: bottomAnchor)
+        ]
+    }
+
+    override open func updateContent() {
+        imageGallery.data = data.flatMap {
+            .init(
+                attachments: $0.attachments.filter { $0.type == .image },
+                didTapOnAttachment: $0.didTapOnAttachment
+            )
+        }
+        imageGallery.isVisible = layoutOptions.contains(.images)
+
+        fileList.data = data.flatMap {
+            .init(
+                attachments: $0.attachments.filter { $0.type == .file },
+                didTapOnAttachment: $0.didTapOnAttachment
+            )
+        }
+        fileList.isVisible = layoutOptions.contains(.files)
+
+        layoutConstraints.values.flatMap { $0 }.forEach { $0.isActive = false }
+        layoutConstraints[layoutOptions]?.forEach { $0.isActive = true }
+    }
+
+    // MARK: - Private
+
+    private var layoutOptions: LayoutOptions {
```

Comment:
This is example of awful computed property.

You iterate `data.attachments` twice inside property, which leads to `O(n)` complexity for property instead of expected `O(1)`.
So in one `updateContent` call you iterate attachments 6 times, for no reason.

Yeah it probably doesn't matter on small amount of data, but unless I missed something, our attachments count is not really limited + we can layout multiple message cells at same time.

Reply:
Absolutely correct! Fixed the `layoutOptions` to iterate over attachments just once and created local variable in `updateContent` in 991cda8125eb4168a56c8be6744670a0f1c08299